### ISSUE TITLE
[prefer-string-starts-ends-with] False positives when slicing

### ISSUE THREAD

**Issue by Zarel** [typescript-eslint/typescript-eslint#2688]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-string-starts-ends-with": ["error"]
  }
}
```

```TS
const filename = 'carpet.txt';
if (filename.slice(0, -4) === 'car') {
	console.log('file is car.txt');
}
```

**Expected Result**

No error

**Actual Result**

```
  2:5  error  Use 'String#startsWith' method instead  @typescript-eslint/prefer-string-starts-ends-with
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.1` |
| `@typescript-eslint/parser`        | `4.4.1` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.11.0` |
| `node`                             | `14.14.0` |

**Comment by bradzacher** [comment]
This just looks like a really cryptic way to do `startsWith` + `length`. Am I correct in saying that's what your code is doing?

**Comment by Zarel** [comment]
I mean, you could also interpret it that way, but I think it's less cryptic than `startsWith` + `length`.

To me, this version is much clearer:

```ts
function isATextFileNamed(filename, name) {
  return filename.endsWith('.txt') && filename.slice(0, -4) === name;
}
```

Than if you wanted to use `startsWith` and `length`:

```ts
function isATextFileNamed(filename, name) {
  return filename.endsWith('.txt') && filename.startsWith(name) && filename.length === name.length + 4;
}
```

**Comment by Zarel** [comment]
Another use-case is if you want to check a first line.

This:

```ts
function firstLineEquals(text, firstLine) {
  const newlinePosition = text.indexOf('\n');
  return text.slice(0, newlinePosition) === firstLine;
}
```

Is much clearer than this:

```ts
function firstLineEquals(text, firstLine) {
  const newlinePosition = text.indexOf('\n');
  return text.startsWith(firstLine) && newlinePosition === firstLine.length;
}
```

But the former also dings `prefer-string-starts-ends-with`.

**Comment by bradzacher** [comment]
> To me, this version is much clearer

IMO it's not because (in my experience) most devs don't have experience with negative index form of `slice`, which means there's non-trivial overhead in understanding that `slice(0, -4)` means "take all but the last 4 characters". TBH if I wasn't paying enough attention, I could easily misread this as `substr(-4)`, which has a very different meaning.

For your second example, I'd use `substring` instead, as I think it communicates the intention more clearly and is easier to read than `slice`.

But this is just bikeshedding.
I was more just double checking I understood your code's intention.

----

The rule doesn't currently understand negative indices in the second position for `slice`.
Happy to accept a PR!
